feat: generic oauth client

barredterra opened the pull request:
Please check #9722 for an explanation.

Based on https://github.com/DigiThinkIT/oauth2_client

### Todo
- [x] Use [Web Application Flow](https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html#web-application-flow) to not reinvent the wheel.
- [x] Integration test against Google
- [x] Integration test against Frappe
- [x] Extract method `update_token_cache(token_cache, token_data)`
- [x] Fix logout on redirect (#11035)
- [x] Accept and save a custom redirect URI in **Token Cache** where the user is sent finally, after storing the access token.
- [x] ~~Add `base_url` to **Connected App** to use it later with rauth_session.get('/oauth/profile').~~ (Not supported by requests_oauthlib.)
- [x] redirect to `#workspace` instead of `#desktop` https://github.com/frappe/frappe/blob/e4928e5bf689d20fc06596ee2070b47f534c0436/frappe/utils/oauth.py#L309
- [x] `access_token` and `refresh_token` should be Password fields.
- [x] Save only the granted scopes.
- [x] Tests
- [x] Add a test for backend application flow
- [x] add a method to get an authenticated requests session.
- [x] Find a solution for non-standardized token revocation. Maybe don't revoke at all?
- [x] Docs
- [x] Separate PR for https://github.com/frappe/frappe/pull/11000/commits/277b082d15976b59cd610efeea997146af8092d8 (#11066)
- [x] Separate PR for https://github.com/frappe/frappe/pull/11000/commits/6a482d152759d73ee1aef3525f2b6c496548207b (#11065)

### Related Issues

#11038
#11039
#11269
#11982

Documentation: https://github.com/frappe/frappe_docs/pull/67
Close #9722

barredterra committed:
feat: connected app and token cache

barredterra committed:
fix: indentation

barredterra committed:
refactor

barredterra committed:
start to use requests_oauthlib

barredterra committed:
fix: redirect to provider

barredterra committed:
add Connected App to desk page

barredterra committed:
OAuth scopes are a table.

Also, show Token Cache in Dashboard of Connected App.

barredterra committed:
use small text to store long tokens

barredterra committed:
open auth flow in a new tab

barredterra committed:
refactor: connected app

barredterra committed:
style: fix sider

surajshetty3416 commented:
@barredterra Can you please explain what problem does this solve?

barredterra commented:
@surajshetty3416 Please see the related issue (#9722). There are at least six different implementations of the same thing right now. The goal is to replace these.

barredterra commented:
@Mangesh-Khairnar, @Thunderbottom in #9969 the maximum password length was fixed to 127. Here I'd like to encrypt access tokens which are longer than that. In Docfield I can increase the legth of password up to 1000. Assuming the encryption adds 13 (`140 - 127`) extra characters, could we set the maximum password length to `docfield.length - 13` ?

barredterra committed:
refactor: token cache updates itself

barredterra committed:
fix: save only new scopes

barredterra committed:
refactor: flatten callback

barredterra committed:
feat: add fields success_uri, token_type

barredterra committed:
feat: return token as json, calculate expiry

barredterra committed:
refactor: connected app

barredterra committed:
refactor: move mothods to token cache

barredterra committed:
fix(oauth provider): parse cookies correctly

barredterra committed:
refactor: extract get_scopes()

barredterra committed:
fix: BackendApplicationClient also likes scopes

barredterra committed:
fix: include client_id when fetching token

barredterra committed:
fix: sider

barredterra committed:
fix(database): allow passwords up to 20000 characters / 60000 bytes (#11039)

barredterra committed:
feat: store tokens as password

barredterra committed:
feat(database): patch for long passswords

barredterra committed:
Merge branch 'develop' into generic_oauth_client

revant commented on frappe/integrations/doctype/token_cache/token_cache.py:
```diff
@@ -0,0 +1,85 @@
+				app.revocation_endpoint,
+				data=urlencode({'token': new_token.get('access_token')}),
+				headers={
+					'Authorization': 'Bearer ' + new_token.get('access_token'),
```
revoke endpoint headers are not standardized.

Some providers use no headers
Some providers use new valid token
Some providers use Basic Auth with client_id as username and client_secret as password.

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra committed:
fat: add field base_url

barredterra committed:
fix: add docstrings

barredterra committed:
fix: redirect to desk#workspace instead of desk#desktop

barredterra committed:
fix: remove client_id from test_records

Because it will be auto-generated and overwritten  anyways.

barredterra committed:
refactor(oauth2): better name and docstring

barredterra committed:
feat: method to create social login key for tests

barredterra committed:
fix: add fallback url for tests

barredterra committed:
fix: commit after saving token

barredterra committed:
fix: add a useless link so that test record gets created

barredterra committed:
tests: add test record

barredterra committed:
tests: web application fow against frappe

revant commented:
There is a field `Useless Link` for DocType User in `Connected App`

works as expected on my [OAuth 2 Authorization Server](https://gitlab.com/castlecraft/building-blocks)

`Token Cache` Screenshot
<details>

![generic_oauth_client_bearer_token](https://user-images.githubusercontent.com/1040161/90539908-16186000-e19e-11ea-9e86-eace0b308489.png)

</details>

barredterra commented:
Thanks for testing, @revant! The useless link is a workaround for #11269.

barredterra committed:
Merge branch 'develop' into generic_oauth_client

revant commented:
If you need local authorization server with docker-compose

refer this https://gitlab.com/castlecraft/building-blocks/-/wikis/docker-compose-for-local-setup

in brief

```
wget https://gitlab.com/castlecraft/building-blocks/-/raw/develop/docker/develop.yaml -O docker-compose.yml
wget https://gitlab.com/castlecraft/building-blocks/-/raw/develop/scripts/setup-wizard
chmod +x setup-wizard
docker-compose up -d
sleep 15
./setup-wizard
```

barredterra committed:
Merge branch 'develop' into generic_oauth_client

rmehta commented:
Much needed @barredterra 👍

barredterra committed:
fix: delete old js test

barredterra committed:
fix: delete old js test

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra committed:
fix: style

barredterra committed:
fix: remove useless link

barredterra committed:
refactor: rename Endpoint to URI

barredterra committed:
fix: remove 'callback' field

barredterra committed:
test: remove unused imports

barredterra committed:
fix(Token Cache): use get_password

barredterra committed:
fix: remove TokenCache.refresh_token()

Will be handled by requests_oauthlib via
ConnectedApp.get_oauth2_session(user).

barredterra committed:
test: authenticated session, backend flow

barredterra committed:
fix: remove base_url

use case not supported by requests, so not needed anymore

sagarvora commented on frappe/utils/oauth.py:
```diff
@@ -306,7 +306,7 @@ def redirect_post_login(desk_user, redirect_to=None, provider=None):
 
 	if not redirect_to:
-		# the #desktop is added to prevent a facebook redirect bug
+		# the #workspace is added to prevent a facebook redirect bug
```
Thanks.

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra committed:
fix: get_auth_header

barredterra committed:
fix: validate token_type, use cstr

barredterra committed:
test: Token Cache

barredterra commented:
Ping

nameduser0 commented:
Looks great Raffael, just wondered if you were planning to add the Token Cache table to the User doctype in the 'Third Party Authentication' section. I guess one day it will replace the 'Social Logins'?

revant commented:
> Token Cache table to the User doctype in the 'Third Party Authentication' section

Token Cache is not a child table, to get python api access to tokens there are wrapper objects created. Once things get finalized the docs will be added. Basic doc was added here https://github.com/DigiThinkIT/oauth2_client/wiki, The api has changed in PR.

> I guess one day it will replace the 'Social Logins'?

All the doctypes which use this data will be linked to this `Connected App` and use its simplified frappe framework api.

nameduser0 commented:
@revant  I understand how it works. The Token Cache (aka Access) table holds keys for Users (and instance) so I thought it would be useful for Users to see what keys they have for what scopes, as they do now, rather than being hidden and removing features that current users might use? I can't see there's anything stopping Token Cache being a child doctype.

And the Connected App table is essentially the same as the Provider table, so not the same as the keys table referenced by the current Social Logins.

barredterra commented:
@casesolved-co-uk I agree that users should be able to see which connected apps they have authorized and be able to revoke access (delete the token). The easiest way would be via a link in the **User** dashboard and appropriate permissions on **Token Cache**.

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra committed:
fix: remove backend application flow

barredterra committed:
fix: Token Cache permissions

barredterra committed:
feat: add Token Cache to User dashboard

barredterra committed:
feat: allow role "All" to view Connected Apps

barredterra committed:
fix: connected app

fix errors from previous refactor

barredterra committed:
fix: token cache has to ignore permissions

barredterra committed:
test: fix connected app

barredterra committed:
fix: remove unused import of BackendApplicationClient

barredterra committed:
fix: TokenCache.get_expires_in()

barredterra committed:
test: fix TestTokenCache.setUp()

barredterra committed:
fix: cint for expires_in, docstring

barredterra committed:
fix: allow insecure transport in tests

revant committed:
fix: Connected App

fix Get OpenID Configuration button
fix permission error on token save

barredterra committed:
Merge pull request #4 from revant/generic_oauth_client

fix: Connected App

nameduser0 commented:
Unstale

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra committed:
fix: allow get_oauth2_session without user arg

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra committed:
fix: show "Connect to {provider}" if doc is not new

barredterra committed:
faet: display provider name in token cache

barredterra committed:
feat: toggle client credential section

barredterra committed:
fix: default if password is empty

barredterra committed:
feat: allow extra parameters

barredterra committed:
fix: redirect back to connected app by default

(instead of to desk)

barredterra committed:
refactor: frappe.db.exists instead of try/except

barredterra committed:
test: fix method name

revant reviewed (dismissed):
Works with my OIDC Authorization server.

LGTM.

barredterra commented:
Thanks for you review, @revant !

barredterra committed:
Merge branch 'develop' into generic_oauth_client

gavindsouza commented:
@barredterra Is there any documentation for this PR yet? If yes, update the link in the PR description too. Also, can you fix the commit messages so that the PR checks pass?

sagarvora commented on frappe/config/integrations.py:
```diff
@@ -77,6 +77,11 @@ def get_data():
 					"name": "OAuth Provider Settings",
 					"description": _("Settings for OAuth Provider"),
 				},
+				{
+					"type": "doctype",
+					"name": "Connected App",
+					"description": _("Connected App"),
+				},
```
I don't know if this is needed. If it is, please ensure description is same as that of the Desk Page.

barredterra replied:
Not sure as well. If it's not, the whole file could be removed, right?

barredterra commented:
@gavindsouza docs are not recognized because of this bug: #11982

barredterra commented:
@gavindsouza fixing published commit messages is [strongly discouraged](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/changing-a-commit-message#amending-older-or-multiple-commit-messages). Maybe this PR can be squashed?

barredterra committed:
fix: description in config

gavindsouza commented:
@barredterra forced pushes are discouraged on all stable branches. However, we're talking about a feature branch of a fork here. I think it's okay. Unless you have any auto deployments linked to this branch that can't handle forced pushes, I don't see any reasons as to why this should be a problem.

We often discourage squashing PRs with such large diffs but maybe @surajshetty3416 can look into this.

barredterra commented:
@gavindsouza @surajshetty3416 I tried to `rebase` and `reword` but wasn't able to because i'd need to touch all changes on `develop` in the last four months. If you know a better way, please let me know.

I think squashing might be okay in this case because it's >90% new files – no big changes to existing files.

barredterra committed:
Merge branch 'develop' into generic_oauth_client

barredterra commented:
Not sure why tests are failing in travis, locally they are working fine. Any Ideas?

```
> bench reinstall
> bench run-tests --module frappe.integrations.doctype.token_cache.test_token_cache
Updating Dashboard for frappe

test_get_auth_header (frappe.integrations.doctype.token_cache.test_token_cache.TestTokenCache) (74.8s)
....
----------------------------------------------------------------------
Ran 4 tests in 75.038s

OK
> bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app

test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL. (8.92s)
.
----------------------------------------------------------------------
Ran 1 test in 8.924s
```

(Webserver with multithreading running)

rmehta commented:
@barredterra why does travis fail?

Edit: sorry did not see your earlier message. Its usually due to the sequence of the tests. Check if there is a commit in a test that may cause this.

revant commented:
`test_token_cache` doesn't fail.

`frappe/integrations/doctype/connected_app/test_connected_app.py` does fail:

```shell
❯ bench new-site test.localhost --mariadb-root-password=`cat /tmp/password` --force

Installing frappe...
Updating DocTypes for frappe        : [========================================] 100%
Updating country info               : [========================================] 100%
*** Scheduler is disabled ***
Current Site set to test.localhost
❯ bench --site test.localhost set-config allow_tests true
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
Updating Dashboard for frappe
Traceback (most recent call last):
  File "/home/revant/.pyenv/versions/3.7.6/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/revant/.pyenv/versions/3.7.6/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/revant/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 99, in <module>
    main()
  File "/home/revant/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 18, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/revant/frappe-bench/env/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/revant/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 26, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/revant/frappe-bench/apps/frappe/frappe/commands/utils.py", line 539, in run_tests
    ui_tests=ui_tests, doctype_list_path=doctype_list_path, failfast=failfast)
  File "/home/revant/frappe-bench/apps/frappe/frappe/test_runner.py", line 73, in main
    ret = run_tests_for_module(module, verbose, tests, profile, junit_xml_output=junit_xml_output)
  File "/home/revant/frappe-bench/apps/frappe/frappe/test_runner.py", line 177, in run_tests_for_module
    make_test_records(doctype, verbose=verbose)
  File "/home/revant/frappe-bench/apps/frappe/frappe/test_runner.py", line 273, in make_test_records
    make_test_records_for_doctype(options, verbose, force)
  File "/home/revant/frappe-bench/apps/frappe/frappe/test_runner.py", line 321, in make_test_records_for_doctype
    frappe.local.test_objects[doctype] += make_test_objects(doctype, test_module.test_records, verbose, force)
  File "/home/revant/frappe-bench/apps/frappe/frappe/test_runner.py", line 372, in make_test_objects
    d.insert()
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/document.py", line 268, in insert
    self.run_post_save_methods()
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/document.py", line 981, in run_post_save_methods
    self.run_method("on_update")
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/document.py", line 848, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/document.py", line 1133, in composer
    return composed(self, method, *args, **kwargs)
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/document.py", line 1116, in runner
    add_to_return_value(self, fn(self, *args, **kwargs))
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/document.py", line 842, in <lambda>
    fn = lambda self, *args, **kwargs: getattr(self, method)(*args, **kwargs)
  File "/home/revant/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 101, in on_update
    self.send_password_notification(self.__new_password)
  File "/home/revant/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 214, in send_password_notification
    self.send_welcome_mail_to_user()
  File "/home/revant/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 271, in send_welcome_mail_to_user
    link = self.reset_password()
  File "/home/revant/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 236, in reset_password
    check_password_reset_limit(self.name, rate_limit)
  File "/home/revant/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 1133, in check_password_reset_limit
    frappe.throw(_("You have reached the hourly limit for generating password reset links. Please try again later."))
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 408, in throw
    msgprint(msg, raise_exception=exc, title=title, indicator='red', is_minimizable=is_minimizable, wide=wide, as_list=as_list)
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 387, in msgprint
    _raise_exception()
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 341, in _raise_exception
    raise raise_exception(msg)
frappe.exceptions.ValidationError: You have reached the hourly limit for generating password reset links. Please try again later.
```

I tried same with `localhost` as site, error is different

```shell
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
Updating Dashboard for frappe
F
======================================================================
FAIL: test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/revant/frappe-bench/apps/frappe/frappe/integrations/doctype/connected_app/test_connected_app.py", line 63, in test_web_application_flow
    self.assertEqual(auth_response.status_code, 200)
AssertionError: 404 != 200

----------------------------------------------------------------------
Ran 1 test in 0.177s

FAILED (failures=1)
```

barredterra commented:
Thanks @revant, i'll try to debug it.

barredterra commented:
Still works for me, @revant 😕 

```
bench --site site1.local run-tests --module frappe.integrations.doctype.connected_app.test_connected_app 

test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL. (2.33e+02s)
.
----------------------------------------------------------------------
Ran 1 test in 232.848s

OK
```

Request log:

```
127.0.0.1 - - [21/Dec/2020 15:21:42] "POST /api/method/login HTTP/1.1" 200 -
127.0.0.1 - - [21/Dec/2020 15:23:43] "GET /api/method/frappe.integrations.oauth2.authorize?response_type=code&client_id=1026c0c2c7&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fapi%2Fmethod%2Ffrappe.integrations.doctype.connected_app.connected_app.callback%2Fc0dee6d87d&scope=all+openid&state=N9s3bM74sHtsq1gUzP6UsKF4rawBce HTTP/1.1" 302 -
127.0.0.1 - - [21/Dec/2020 15:23:43] "GET /api/method/frappe.integrations.oauth2.approve?response_type=code&client_id=1026c0c2c7&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fapi%2Fmethod%2Ffrappe.integrations.doctype.connected_app.connected_app.callback%2Fc0dee6d87d&scope=all%20openid&state=N9s3bM74sHtsq1gUzP6UsKF4rawBce HTTP/1.1" 302 -
/home/frappe/frappe-bench/env/lib/python3.6/site-packages/bs4/__init__.py:389: UserWarning: "http://localhost:8000/api/method/frappe.integrations.doctype.connected_app.connected_app.callback/c0dee6d87d" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  ' that document to Beautiful Soup.' % decoded_markup
127.0.0.1 - - [21/Dec/2020 15:23:53] "POST /api/method/frappe.integrations.oauth2.get_token HTTP/1.1" 200 -
127.0.0.1 - - [21/Dec/2020 15:23:53] "GET /api/method/frappe.integrations.doctype.connected_app.connected_app.callback/c0dee6d87d?code=MALHmicblsuoi4snYllINQjPpDU017&state=N9s3bM74sHtsq1gUzP6UsKF4rawBce HTTP/1.1" 302 -
127.0.0.1 - - [21/Dec/2020 15:23:57] "GET /desk HTTP/1.1" 200 -
127.0.0.1 - - [21/Dec/2020 15:24:36] "GET /desk HTTP/1.1" 200 -
127.0.0.1 - - [21/Dec/2020 15:25:10] "POST /api/method/frappe.integrations.oauth2.get_token HTTP/1.1" 200 -
127.0.0.1 - - [21/Dec/2020 15:25:10] "GET /api/method/frappe.auth.get_logged_user HTTP/1.1" 200 -
```

Tested commit: 689e39a

barredterra committed:
fix: base_url

barredterra committed:
fix: test records

barredterra commented:
I noticed that `run-test --module` and `run-tests` result in different redirect URIs. With `run-tests` the current request URL when Connected App gets created is `'http://localhost'` and with `run-test --module` it is `None`. Tried to fix that, works locally, let's see what travis says.

revant commented:
I had `"maintenance_mode": 1` and `"pause_scheduler": 1` in common_site_config.json, 503 was because of that.
I remove that and tried again. Now the result is same as travis.

```
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
Updating Dashboard for frappe
E
======================================================================
ERROR: test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/revant/frappe-bench/apps/frappe/frappe/integrations/doctype/connected_app/test_connected_app.py", line 69, in test_web_application_flow
    token = token_cache.get_password('access_token')
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/base_document.py", line 757, in get_password
    return get_decrypted_password(self.doctype, self.name, fieldname, raise_exception=raise_exception)
  File "/home/revant/frappe-bench/apps/frappe/frappe/utils/password.py", line 51, in get_decrypted_password
    frappe.throw(_('Password not found'), frappe.AuthenticationError)
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 408, in throw
    msgprint(msg, raise_exception=exc, title=title, indicator='red', is_minimizable=is_minimizable, wide=wide, as_list=as_list)
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 387, in msgprint
    _raise_exception()
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 341, in _raise_exception
    raise raise_exception(msg)
frappe.exceptions.AuthenticationError: Password not found

----------------------------------------------------------------------
Ran 1 test in 2.059s

FAILED (errors=1)
```

barredterra committed:
fix: rely on frappe.utils.get_url

revant commented:
still the same error locally

```shell
❯ bench drop-site localhost --no-backup --root-password=`cat /tmp/password`
❯ cd apps/frappe
❯ bench new-site localhost --mariadb-root-password=`cat /tmp/password` --force

Installing frappe...
Updating DocTypes for frappe        : [========================================] 100%
Updating country info               : [========================================] 100%
*** Scheduler is disabled ***
Current Site set to localhost
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
Testing is disabled for the site!
You can enable tests by entering following command:
bench --site localhost set-config allow_tests true
❯ bench --site localhost set-config allow_tests true
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
Updating Dashboard for frappe
E
======================================================================
ERROR: test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/revant/frappe-bench/apps/frappe/frappe/integrations/doctype/connected_app/test_connected_app.py", line 69, in test_web_application_flow
    token = token_cache.get_password('access_token')
  File "/home/revant/frappe-bench/apps/frappe/frappe/model/base_document.py", line 757, in get_password
    return get_decrypted_password(self.doctype, self.name, fieldname, raise_exception=raise_exception)
  File "/home/revant/frappe-bench/apps/frappe/frappe/utils/password.py", line 51, in get_decrypted_password
    frappe.throw(_('Password not found'), frappe.AuthenticationError)
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 408, in throw
    msgprint(msg, raise_exception=exc, title=title, indicator='red', is_minimizable=is_minimizable, wide=wide, as_list=as_list)
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 387, in msgprint
    _raise_exception()
  File "/home/revant/frappe-bench/apps/frappe/frappe/__init__.py", line 341, in _raise_exception
    raise raise_exception(msg)
frappe.exceptions.AuthenticationError: Password not found

----------------------------------------------------------------------
Ran 1 test in 2.354s

FAILED (errors=1)
❯ cd apps/frappe
❯ git status
On branch generic_oauth_client
Your branch is up to date with 'alyf-de/generic_oauth_client'.

nothing to commit, working tree clean
```

branch on latest

```shell
❯ git status
On branch generic_oauth_client
Your branch is up to date with 'alyf-de/generic_oauth_client'.

nothing to commit, working tree clean

❯ git log

commit 301ed4c7e322fb1de93a197ee3423992d8719435 (HEAD -> generic_oauth_client, alyf-de/generic_oauth_client)
Author: barredterra <14891507+barredterra@users.noreply.github.com>
Date:   Tue Dec 22 18:35:26 2020 +0100

    fix: rely on frappe.utils.get_url
```

barredterra committed:
feat: redirect to Guest to login

barredterra committed:
refactor: useless try except, better error messages

revant commented:
I changed the `oauth_client/test_records.json` : 

changed `localhost` to `localhost:8000`

```json
[
 {
  "app_name": "_Test OAuth Client",
  "client_secret": "test_client_secret", 
  "default_redirect_uri": "http://localhost:8000", 
  "docstatus": 0, 
  "doctype": "OAuth Client", 
  "grant_type": "Authorization Code", 
  "name": "test_client_id", 
  "redirect_uris": "http://localhost:8000", 
  "response_type": "Code", 
  "scopes": "all openid", 
  "skip_authorization": 1
 }
]
```

test worked:

```shell
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app

test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL. (3.74s)
.
----------------------------------------------------------------------
Ran 1 test in 3.735s

OK
❯ bench --site localhost reinstall --mariadb-root-password `cat /tmp/password` --yes --admin-password admin

Installing frappe...
Updating DocTypes for frappe        : [========================================] 100%
Updating country info               : [========================================] 100%
*** Scheduler is disabled ***
❯ bench run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
Updating Dashboard for frappe

test_web_application_flow (frappe.integrations.doctype.connected_app.test_connected_app.TestConnectedApp)
Simulate a logged in user who opens the authorization URL. (3.77s)
.
----------------------------------------------------------------------
Ran 1 test in 3.767s

OK
```

tests worked

barredterra commented:
Here's a (hopefully) reproducible example:

```bash
bench --version # 5.0.0
bench init oauth-test
cd oauth-test/apps/frappe
git remote add origin https://github.com/alyf-de/frappe
git fetch origin
git checkout --track origin/generic_oauth_client
cd ../..
bench new-site oauth-test
bench --site oauth-test set-config developer_mode 1 # needs developer mode because frappe.flags.in_test is not set in web callback.
bench --site oauth-test set-config allow_tests true
bench --site oauth-test add-to-hosts
bench start
bench --site oauth-test run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
# First test-run fails with
# requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))

# restart bench
bench --site oauth-test run-tests --module frappe.integrations.doctype.connected_app.test_connected_app
# Second test-run works
```

Errors identified:

- Tests depend on developer mode. If developer mode is not enabled, `connected_app.callback` raises `oauthlib.oauth2.rfc6749.errors.InsecureTransportError`.
- The first test-run fails, the second (and all after that) works.

barredterra committed:
fix: allow insecure transport in travis

barredterra commented:
@rmehta tests are passing now :)

revant commented:
Tested manually. Works as expected.

barredterra commented:
Any other changes required, @surajshetty3416?
